### PR TITLE
[rtsan] Make sure rtsan gets initialized on mac

### PR DESCRIPTION
**Pull request by cjappl**
Intermittently on my mac I was getting the same nullptr crash in dlsym.

We need to make sure rtsan gets initialized on mac between when the binary starts running, and the first intercepted function is called. Until that point we should use the DlsymAllocator.

### CODE REVIEW

**Comment by cjappl** [comment]
CC For review @vitalybuka @MaskRay @davidtrevelyan

**Comment by cjappl** [comment]
Pinging reviewers @MaskRay @vitalybuka 

Thanks as always for the help.

File: compiler-rt/lib/rtsan/rtsan_interceptors.cpp
Comment (vitalybuka) [inline compiler-rt/lib/rtsan/rtsan_interceptors.cpp:42]: It would be better to use atomics here
see `static atomic_uint8_t asan_inited`

Comment (cjappl) [reply compiler-rt/lib/rtsan/rtsan_interceptors.cpp]: Fixed in this latest version. Thanks!

**Comment by cjappl** [comment]
Weekly ping of reviewers: @vitalybuka @MaskRay 

Current state: All outstanding review comments have been addressed.

File: compiler-rt/lib/rtsan/rtsan.cpp
Comment (vitalybuka) [inline compiler-rt/lib/rtsan/rtsan.cpp:39]: why don't you call a static one here?

File: compiler-rt/lib/rtsan/rtsan.cpp
Comment (vitalybuka) [inline compiler-rt/lib/rtsan/rtsan.cpp:38]: I guess it's cleaner with a mutex, like `asan_inited_mutex`

There is possibility that 2 racing thread will pass `CHECK(!IsInitRunning());`

File: compiler-rt/lib/rtsan/rtsan.cpp
Comment (cjappl) [reply compiler-rt/lib/rtsan/rtsan.cpp:39]: Is there a benefit of calling a static function vs the external API function?

Comment (cjappl) [reply compiler-rt/lib/rtsan/rtsan.cpp:39]: (for the record, the external method must exist because we use it in rtsan_interceptors to see if we need to use the start-up allocator before dlsym is fully settled)

File: compiler-rt/lib/rtsan/rtsan.cpp
Comment (cjappl) [reply compiler-rt/lib/rtsan/rtsan.cpp]: Good catch. Alright, this now does the following:

Checks if initialized - using an atomic to prevent data races
If not, grabs the mutex
Once the mutex is grabbed, checks the atomic once more to ensure we don't double initialize.

I think this handles all the possible data races in this path.

The separation of rtsan_ensure_inited vs rtsan_init is to make sure the preinit array initialization of rtsan is as lean as possible. This is following some of the patterns I saw in other sanitizers.

**Review by vitalybuka — APPROVED** [review]

**Comment by cjappl** [comment]
Two very small updates since last review:

1. Matching llvm style "If you use a braced initializer list when initializing a variable, use an equals before the open curly brace:"
```cpp
- static atomic_uint8_t rtsan_initialized{0};
+ static atomic_uint8_t rtsan_initialized = {0};
```

2. Got rid of the now-unused `IsInitialized` in favor of the `__rtsan_is_initialized` method, this was cruft from an earlier version of this review.

These two changes may be reviewed in the latest two commits, if anyone wants to see them in isolation.

I also rebased on latest main. I will likely merge this early next week

**Comment by cjappl** [comment]
Thanks for the merge @vitalybuka !

I went ahead and requested commit access to LLVM, so I'll be able to merge things next time I get approvals. I hope that leads to less clerical tasks on your plate.

**Pull request merged**
